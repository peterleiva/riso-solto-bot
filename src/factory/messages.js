/** @typedef {string[]} Dataset */
/** @typedef {[{text: string, type: 'text'}]} Messages */

import { readFile } from "fs/promises";
import createDebug from "debug";
import { fileURLToPath } from "url";
import { dirname, resolve } from "path";
import Text from "./text.js";

const __dirname = dirname(fileURLToPath(import.meta.url));
const debug = createDebug("factory:message");
const DATASET_PATH = resolve(__dirname, "../dataset.json");

/**
 *
 * @return {Dataset}
 */
async function loader() {
  try {
    const content = await readFile(DATASET_PATH, "utf-8");
    debug("dataset read %s", DATASET_PATH);

    const dataset = JSON.parse(content);
    debug("dataset: %O", dataset);

    return dataset;
  } catch (error) {
    console.error("Couldn't load src/dataset");
    console.error(error);
  }
}

/**
 * Message factory, creates message tokens
 *
 * @returns {{create: () => Promise<Messages>}}
 */
export function messageFactory() {
  let dataset;

  /**
   * @param {Dataset}
   * @return {Messages}
   */
  function transform(dataset) {
    return dataset.map(data => new Text({ text: data }));
  }

  return {
    /**
     *
     * @return {Promise<Messages>}
     */
    create: async () => {
      if (!dataset) {
        dataset = transform(await loader());
      }

      return dataset;
    },
  };
}
