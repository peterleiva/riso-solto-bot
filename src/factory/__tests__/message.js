import { messageFactory } from "../messages.js";

test("creates message factory", async () => {
  const factory = messageFactory();
  const dataset = await factory.create();

  expect(dataset).toBeInstanceOf(Array);
  expect(dataset[0]).toHaveProperty("type", "text");
  expect(dataset[0]).toHaveProperty("text");

  expect(dataset).toMatchSnapshot();
});
