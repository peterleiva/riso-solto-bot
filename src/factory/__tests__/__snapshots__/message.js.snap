// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates message factory 1`] = `
Array [
  Object {
    "text": "😀",
    "type": "text",
  },
  Object {
    "text": "😃",
    "type": "text",
  },
  Object {
    "text": "😄",
    "type": "text",
  },
  Object {
    "text": "😁",
    "type": "text",
  },
  Object {
    "text": "😆",
    "type": "text",
  },
  Object {
    "text": "😂",
    "type": "text",
  },
  Object {
    "text": "🤣",
    "type": "text",
  },
  Object {
    "text": "KKKK",
    "type": "text",
  },
  Object {
    "text": "haha",
    "type": "text",
  },
  Object {
    "text": "rsrs",
    "type": "text",
  },
  Object {
    "text": "jajaja",
    "type": "text",
  },
  Object {
    "text": "lol",
    "type": "text",
  },
  Object {
    "text": "lmfao",
    "type": "text",
  },
  Object {
    "text": "lmfout",
    "type": "text",
  },
  Object {
    "text": "asdjiasjdihaus",
    "type": "text",
  },
]
`;
